### ISSUE TITLE
Support for multiple languages in file

### ISSUE THREAD

**Issue by zulus** [dschaefer/antlng#2]
Have you plan to support multiple languages in file? Important for WEB.  Examples:
html file, possible mixins:
- javascript
- css

ant xml file, possible mixins:
- javascript
- ruby
- what ever jvm language

java. possible mixins:
- SQL
- HQL

Off course always should be main content type (html, xml,php, java), and ability to report embedded.

**Comment by dschaefer** [comment]
Yes, definitely. This is one of the features of Antlr 4 which is pretty appealing. Not sure of the details yet, but things like JSX used by React.js which include XML annotation inside a JavaScript file is one of the things I'm looking for. The HTML example you bring up is a natural as well and whatever else we put together.

**Comment by zulus** [comment]
So should look at him closer. I has hope that JavaCup2 will introduce extendable parsers but project was abandoned :( 

Based on description You plan fix bug 42046 (http://eclip.se/42046) and maybe also bug 225994 (http://eclip.se/225994) 

Anyway eclipse have filename centric content types and this might be a barrier

**Comment by dschaefer** [comment]
I think the key is to capture the island language (using Antlr terms) in a token. Then pass it off to another lexer/parser based in context information to finish the job.

**Comment by zulus** [comment]
This can be done if you know where should be injected (like css/javascript in html/xml). But what with languages that can be anywhere:
1. Tamplate languages like PHP/JSP which can be injected anywhere (php:<?php ?>; twig: {{ $varname }} ). They always have START sequence, and sometimes END sequence.
2. SQL/HQL/HTML in string literal, it's not always simple to detect them.

I thinking about extendable highlighter/parser/model since my first day with eclipse text editor :P

**Comment by dschaefer** [comment]
You'd have to replicate how the underlying runtimes figure it out.
